kms-message/src/kms_crypto_apple.c

```diff
@@ -58,4 +61,83 @@ kms_sha256_hmac (void *unused_ctx,
    return true;
 }
 
+static void
+safe_CFRelease (CFTypeRef ptr)
+{
+   if (ptr) {
+      CFRelease (ptr);
+   }
+}
+
+bool
+kms_sign_rsaes_pkcs1_v1_5 (void *unused_ctx,
+                           const char *private_key,
+                           size_t private_key_len,
+                           const char *input,
+                           size_t input_len,
+                           unsigned char *signature_out)
+{
+   CFDataRef key_data_ref = NULL;
+   CFDataRef pass_ref = NULL;
+   SecItemImportExportKeyParameters import_params;
+   OSStatus status;
+   /* TODO: I think the expected format should be kSecFormatWrappedPKCS8, but
+    * GCP keys appear to only load for kSecFormatBSAFE. */
+   SecExternalFormat format = kSecFormatUnknown;
+   SecExternalItemType type = kSecItemTypePrivateKey;
+   CFArrayRef out_ref = NULL;
+   SecKeyRef key_ref = NULL;
+   CFDataRef data_to_sign_ref = NULL;
+   CFErrorRef error_ref;
+   CFDataRef signature_ref = NULL;
+   bool ret = false;
+
+   key_data_ref = CFDataCreate (NULL /* default allocator */,
+                                (const uint8_t *) private_key,
+                                (CFIndex) private_key_len);
+   memset (&import_params, 0, sizeof (SecItemImportExportKeyParameters));
+   import_params.version = SEC_KEY_IMPORT_EXPORT_PARAMS_VERSION;
+
+   /* Give an empty password. SecItemImport returns an error expecting a
+    * password. */
+   pass_ref = CFDataCreate (NULL, NULL, 0);
+   import_params.passphrase = (CFTypeRef) pass_ref;
+
+   status = SecItemImport (key_data_ref,
+                           NULL /* extension. */,
+                           &format,
+                           &type,
+                           0,
+                           &import_params,
+                           NULL /* keychain */,
+                           &out_ref);
+   if (status != errSecSuccess) {
+      goto cleanup;
+   }
+   if (1 != CFArrayGetCount (out_ref)) {
+      goto cleanup;
+   }
+
+   key_ref = (SecKeyRef) CFArrayGetValueAtIndex (out_ref, 0);
+   data_to_sign_ref = CFDataCreate (NULL, (const uint8_t *) input, input_len);
+   error_ref = NULL;
+   signature_ref =
+      SecKeyCreateSignature (key_ref,
+                             kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256,
```

Comment:
Aside, `kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256` is a distinct algorithm (and produces different results). I'm not sure what the difference is (my best guess is that it is the order of signing / hashing).